ext/MitsubaLoader/CMitsubaLoader.cpp

```diff
@@ -1197,14 +1213,49 @@ reg_t registers[REG_COUNT];
 void setCurrBSDFParams(in vec3 n, in vec3 L)
 {
 	vec3 campos = irr_glsl_SBasicViewParameters_GetEyePos(CamData.params.NormalMatAndEyePos);
-	irr_glsl_ViewSurfaceInteraction interaction = irr_glsl_calcFragmentShaderSurfaceInteraction(campos, WorldPos, n);
+	irr_glsl_IsotropicViewSurfaceInteraction interaction = irr_glsl_calcFragmentShaderSurfaceInteraction(campos, WorldPos, n);
 	irr_glsl_BSDFIsotropicParams isoparams = irr_glsl_calcBSDFIsotropicParams(interaction, L);
 	//TODO: T,B tangents
 	vec3 T = vec3(1.0,0.0,0.0);
 	vec3 B = vec3(0.0,0.0,1.0);
 	currBSDFParams = irr_glsl_calcBSDFAnisotropicParams(isoparams, T, B);
 }
 
+vec2 getAlpha(in uint op, in uint ndf, in instr_t instr, in bsdf_data_t data, in mat2 dUV)
+{
+	vec2 a;
+	if (op==OP_ROUGHDIFFUSE)
+	{
+		vec3 scale = decodeRGB19E7(data.data[1].zw);
+		float a_ = textureOrF32(data.data[0].xy, instr_getAlphaUTexPresence(instr), dUV)*scale.x;
+		a = vec2(a_);
+	}
+	else if (op==OP_ROUGHDIELECTRIC)
+	{
+		vec3 scale = decodeRGB19E7(uvec2(data.data[1].w, data.data[2].x));
+		a = vec2( textureOrF32(data.data[0].yz, instr_getAlphaUTexPresence(instr), dUV)*scale.x );
+		if (ndf == NDF_AS)
+			a.y = textureOrF32(uvec2(data.data[0].w, data.data[1].x), instr_getAlphaVTexPresence(instr), dUV)*scale.y;
+	}
+	else if (op==OP_ROUGHCONDUCTOR)
+	{
+		vec3 scale = decodeRGB19E7(data.data[2].zw);
+		a = vec2( textureOrF32(data.data[0].xy, instr_getAlphaUTexPresence(instr), dUV)*scale.x );
+		if (ndf==NDF_AS)
+			a.y = textureOrF32(data.data[0].zw, instr_getAlphaVTexPresence(instr), dUV)*scale.y;
+	}
+	else if (op==OP_ROUGHPLASTIC)
+	{
+		vec3 scale = decodeRGB19E7(uvec2(data.data[1].w, data.data[2].x));
+		a = vec2( textureOrF32(data.data[0].yz, instr_getAlphaUTexPresence(instr), dUV)*scale.x );
+	}
+	else if (op==OP_ROUGHCOATING)
+	{
+		//later
+	}
+	return a;
+}
+
 void instr_execute_DIFFUSE(in instr_t instr, in uvec3 regs, in mat2 dUV, in bsdf_data_t data)
 {
 	if (currBSDFParams.isotropic.NdotL>FLT_MIN)
```

Comment:
I'd move this `if (currBSDFParams.isotropic.NdotL>FLT_MIN)` outside the divergence and all the way to the `else` of `twosided` in the calling function, this way we can skip the entire intruction executing function for anything that is not roughdielectric or difftrans

Reply:
basically pair it with a `isBRDF` function (BRDF cannot be lit or viewed from the back, unless ofc `twosided`, but then you can only light it from the side you can view it, a.k.a. NdotV and NdotL need to be `abs()`>FLT_MIN have the same sign, plus `NdotV>FLT_MIN || twosided`)

Reply:
let me know if tackled in new implementation.

Reply:
?